src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/security/ApplicationUserPrincipal.java

```diff
@@ -0,0 +1,62 @@
+package com.project.portfolio.springboot.onlinesellingandshopping.serverapp.security;
+import com.project.portfolio.springboot.onlinesellingandshopping.serverapp.module.AuthGroup;
+import com.project.portfolio.springboot.onlinesellingandshopping.serverapp.module.User;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+
+import java.util.*;
+
+public class ApplicationUserPrincipal implements UserDetails {
+
+    private final User user;
+    private final List<AuthGroup> authGroups;
+
+    public ApplicationUserPrincipal(User user, List<AuthGroup> authGroups) {
+        super();
+        this.user = user;
+        this.authGroups = authGroups;
+    }
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
```

Comment:
If I understand this logic correctly, it looks like grantedAuthorities depend only on authGroups, in this case you can make this calculation once in the constructor, make it a final instance variable ( class field ), and retrieve the same collection every time. Creating a new collection on every request can create an unnecessary Java memory overhead / make GarbageCollection work overtime.